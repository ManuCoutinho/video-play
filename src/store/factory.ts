import { useStore as store } from '.'
import { baseUrl } from '@/constants/baseUrl'

export async function loadInitialData() {
  store.setState({ isLoading: true })
  const controller = new AbortController()
  const signal = controller.signal
  try {
    await fetch(`${baseUrl}/courses/1`, { signal })
      .then((res) => res.json())
      .then((data) =>
        store.setState({
          course: data,
          isLoading: false
        })
      )
  } catch (err) {
    console.error(err)
    //handle error
  } finally {
    controller.abort()
  }
}

export function playNextVideo() {
  const { currentLessonIndex, currentModuleIndex, course } = store.getState()
  const nextLessonIndex = currentLessonIndex + 1
  const nextLesson =
    course?.modules[currentModuleIndex].lessons[nextLessonIndex]

  if (nextLesson) {
    store.setState({
      currentLessonIndex: nextLessonIndex
    })
  } else {
    const nextModuleIndex = currentModuleIndex + 1
    const nextModule = course?.modules[nextModuleIndex]
    if (nextModule) {
      store.setState({
        currentModuleIndex: nextModuleIndex,
        currentLessonIndex: 0
      })
    }
  }
}

export function onPlay(payload: [number, number]) {
  const [moduleIndex, lessonIndex] = payload
  store.setState({
    currentLessonIndex: lessonIndex,
    currentModuleIndex: moduleIndex
  })
}
