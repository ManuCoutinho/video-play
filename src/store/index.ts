import { create } from 'zustand'
import { PlayerStateType } from '@/models/player'
import * as factory from './factory'

type MethodsPlayer = {
  play: (payload: [number, number]) => void
  next: () => void
  load: () => Promise<void>
}
const initialState: PlayerStateType = {
  course: null,
  currentLessonIndex: 0,
  currentModuleIndex: 0,
  isLoading: true
}

const useStore = create<PlayerStateType & MethodsPlayer>(() => {
  return {
    ...initialState,
    play: (payload: [number, number]) => factory.onPlay(payload),
    next: () => factory.playNextVideo(),
    load: () => factory.loadInitialData()
  }
})

export { useStore }
