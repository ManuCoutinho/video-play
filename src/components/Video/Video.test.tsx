import '@testing-library/jest-dom'
import { screen, render } from '@testing-library/react'
import { describe, it } from 'vitest'
import { Video } from '.'
import { state } from '@/__mocks__/data/playerState'
import { useStore as store } from '@/store'
import { act } from 'react-dom/test-utils'

const mockInitialState = store.getState()

describe('<Video/>', () => {
  beforeEach(() => beforeEach(() => store.setState(mockInitialState)))
  it('should render Video before load data ', () => {
    render(<Video />)

    expect(screen.getByRole('status', { name: /loading/i })).toBeInTheDocument()
    expect(screen.queryByTestId('player')).not.toBeInTheDocument()
  })
  it('should render Video with global data', () => {
    render(<Video />)
    act(() => {
      store.setState(state)
    })
    expect(screen.getByTestId('player')).toBeInTheDocument()
  })
  it('should match to snapshot', () => {
    const { container } = render(<Video />)
    expect(container).toMatchSnapshot()
  })
})
