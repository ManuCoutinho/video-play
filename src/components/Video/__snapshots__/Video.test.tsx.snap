// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Video/> > should match to snapshot 1`] = `
<div>
  <div
    aria-controls="player"
    class="w-full bg-night aspect-video"
  >
    <div
      aria-label="loading"
      aria-live="polite"
      class="flex h-full items-center justify-center"
      role="status"
    >
      <svg
        class="w-6 h-6 animate-spin"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M136,32V64a8,8,0,0,1-16,0V32a8,8,0,0,1,16,0Zm88,88H192a8,8,0,0,0,0,16h32a8,8,0,0,0,0-16Zm-45.09,47.6a8,8,0,0,0-11.31,11.31l22.62,22.63a8,8,0,0,0,11.32-11.32ZM128,184a8,8,0,0,0-8,8v32a8,8,0,0,0,16,0V192A8,8,0,0,0,128,184ZM77.09,167.6,54.46,190.22a8,8,0,0,0,11.32,11.32L88.4,178.91A8,8,0,0,0,77.09,167.6ZM72,128a8,8,0,0,0-8-8H32a8,8,0,0,0,0,16H64A8,8,0,0,0,72,128ZM65.78,54.46A8,8,0,0,0,54.46,65.78L77.09,88.4A8,8,0,0,0,88.4,77.09Z"
        />
      </svg>
    </div>
  </div>
</div>
`;
